ethstore/src/dir/vault.rs

```diff
@@ -0,0 +1,416 @@
+		// check that vault directory does not exists
+		let vault_dir_path = make_vault_dir_path(root, name, true)?;
+		if vault_dir_path.exists() {
+			return Err(Error::CreationFailed);
+		}
+
+		// create vault && vault file
+		fs::create_dir_all(&vault_dir_path)?;
+		if let Err(err) = create_vault_file(&vault_dir_path, &key) {
+			let _ = fs::remove_dir_all(&vault_dir_path); // can't do anything with this
+			return Err(err);
+		}
+
+		Ok(DiskDirectory::new(vault_dir_path, VaultKeyFileManager::new(key)))
+	}
+
+	pub fn at<P>(root: P, name: &str, key: VaultKey) -> Result<Self, Error> where P: AsRef<Path> {
+		// check that vault directory exists
+		let vault_dir_path = make_vault_dir_path(root, name, true)?;
+		if !vault_dir_path.is_dir() {
+			return Err(Error::CreationFailed);
+		}
+
+		// check that passed key matches vault file
+		check_vault_file(&vault_dir_path, &key)?;
+
+		Ok(DiskDirectory::new(vault_dir_path, VaultKeyFileManager::new(key)))
+	}
+
+	fn create_temp_vault(&self, key: VaultKey) -> Result<VaultDiskDirectory, Error> {
+		let original_path = self.path().expect("self is instance of DiskDirectory; DiskDirectory always returns path; qed");
+		let mut path: PathBuf = original_path.clone();
+		let name = self.name();
+
+		path.push(name); // to jump to the next level
+
+		let mut index = 0;
+		loop {
+			let name = format!("{}_temp_{}", name, index);
+			path.set_file_name(&name);
+			if !path.exists() {
+				return VaultDiskDirectory::create(original_path, &name, key);
+			}
+
+			index += 1;
+		}
+	}
+
+	fn copy_to_vault(&self, vault: &VaultDiskDirectory, vault_key: &VaultKey) -> Result<(), Error> {
+		let password = &self.key_manager().key.password;
+		for account in self.load()? {
+			let filename = account.filename.clone().expect("self is instance of DiskDirectory; DiskDirectory fills filename in load; qed");
+			let new_account = account.change_password(password, &vault_key.password, vault_key.iterations)?;
+			vault.insert_with_filename(new_account, filename)?;
+		}
+
+		Ok(())
+	}
+
+	fn delete(&self) -> Result<(), Error> {
+		let path = self.path().expect("self is instance of DiskDirectory; DiskDirectory always returns path; qed");
+		fs::remove_dir_all(path).map_err(Into::into)
+	}
+}
+
+impl VaultKeyDirectory for VaultDiskDirectory {
+	fn as_key_directory(&self) -> &KeyDirectory {
+		self
+	}
+
+	fn name(&self) -> &str {
+		self.path()
+			.expect("self is instance of DiskDirectory; DiskDirectory always returns path; qed")
+			.file_name()
+			.expect("last component of path is checked in make_vault_dir_path; it contains no fs-specific characters; file_name only returns None if last component is fs-specific; qed")
+			.to_str()
+			.expect("last component of path is checked in make_vault_dir_path; it contains only valid unicode characters; to_str fails when file_name is not valid unicode; qed")
+	}
+
+	fn set_key(&self, key: VaultKey, new_key: VaultKey) -> Result<(), SetKeyError> {
```

Comment:
As there's no way to make atomic cross-platform swap of two directories, here's my attempt to minimize potential losses:
1) create temp vault
2) copy all accounts to the temp vault [with same file names]
3) move all files from temp vault to self [with replace] - if crash/shutdown occurs here, directory is partially moved => user can finish this move.
I also thought on some making some '.dirty' file when vault is in inconsistent state && then automatically continue moving on next opening.
But I think it is too much && it could corrupt vault totally.